### PR TITLE
Parentheses readability

### LINKED ISSUE #2546: Update matchers formatting to be more readable
I think we should be formatting using tabs and whitespace more liberally for this usecase. The current syntax is nearly impossible to read.

### PR DESCRIPTION
**Pull request by eddieliao**
Add a cppcheck rule for four or more consecutive parentheses; closes #2546.

### CODE REVIEW

**Comment by causten** [comment]
Can you add a before/after screenshot

**Comment by CharlieL7** [comment]
We probably want to fix the current warnings in this PR also. I can make some commits if it's not clear how to fix them

**Comment by pfultz2** [comment]
There are a lot of warning outside of matchers, which I think a limit of 4 is too aggressive. We could just target the check to the `matcher()` functions with something like this:

```python
@cppcheck.checker
def MatcherNestedParentheses(cfg, data):
    for token in cfg.tokenlist:
        if not simpleMatch(token, 'matcher ( ) const {'):
            continue
        for tok2 in token.tokAt(4, token.linkAt(4)):
            if not simpleMatch(tok2, ") ) ) )"):
                continue
            cppcheck.reportError(
                tok2, "style",
                "Too many nested parentheses can affect readability; consider using variables instead."
            )
            break
```

**Comment by eddieliao** [comment]
> Can you add a before/after screenshot

What screenshots would you like here?

> We probably want to fix the current warnings in this PR also. I can make some commits if it's not clear how to fix them

That would be appreciated; just an example or two would help.



> There are a lot of warning outside of matchers, which I think a limit of 4 is too aggressive. We could just target the check to the `matcher()` functions with something like this:
> 
> ```python
> @cppcheck.checker
> def MatcherNestedParentheses(cfg, data):
>     for token in cfg.tokenlist:
>         if not simpleMatch(token, 'matcher ( ) const {'):
>             continue
>         for tok2 in token.tokAt(4, token.linkAt(4)):
>             if not simpleMatch(tok2, ") ) ) )"):
>                 continue
>             cppcheck.reportError(
>                 tok2, "style",
>                 "Too many nested parentheses can affect readability; consider using variables instead."
>             )
>             break
> ```

4 is pretty aggressive I agree, although I was just basing this off of the original comment of "more than three parentheses". I can go ahead and use your suggestion to only target `matcher()`; let me know if there's another number you have in mind for consecutive parentheses.

**Comment by CharlieL7** [comment]
Fixed the fuse_concat matcher syntax. The other warnings will disappear with the cppcheck change.

File: tools/cppcheck/migraphx.py [hunk 1]
```diff
@@ -429,11 +429,15 @@ def useStlAlgorithms(cfg, data):
...
+def MatcherNestedParentheses(cfg, data):
     for token in cfg.tokenlist:
-        if not simpleMatch(token, ") ) ) )"):
+        if not simpleMatch(token, 'matcher ( ) const {'):
             continue
-        cppcheck.reportError(
-            token, "style",
-            "Too many nested parentheses can affect readability; consider using variables instead."
-        )
+        for tok2 in token.tokAt(4, token.linkAt(4)):
```
Comment (pfultz2) [inline tools/cppcheck/migraphx.py]: You need to do `token.tokAt(4).forward(token.linkAt(4))`. This is most likely where the python error is coming from.

File: tools/cppcheck/migraphx.py [hunk 2]
```diff
@@ -431,9 +431,11 @@ def useStlAlgorithms(cfg, data):
 @cppcheck.checker
 def MatcherNestedParentheses(cfg, data):
     for token in cfg.tokenlist:
-        if not simpleMatch(token, 'matcher ( ) const {'):
+        if not simpleMatch(token, "matcher ( ) const {"):
             continue
-        for tok2 in token.tokAt(4, token.linkAt(4)):
+        for tok2 in cfg.tokenlist[cfg.tokenlist.index(token):]:
```
Comment (pfultz2) [inline tools/cppcheck/migraphx.py:436]: It better to just use the `forward` method to iterate the tokens lazily:

https://github.com/danmar/cppcheck/blob/main/addons/cppcheckdata.py#L503

Comment (eddieliao) [reply tools/cppcheck/migraphx.py]: Changed, thanks

**Review by CharlieL7 — APPROVED** [review]

**Pull request merged**